### PR TITLE
WIP: Refactor accessors, unify usage, make "recipe"

### PR DESCRIPTION
**Pull request by jbrockmendel**
xref #14781, #16890

There are currently accessors for `dt`, `str`, `cat`, and `plot`.  These are each constructed in different ways and using slightly different naming conventions, turning a fairly simple underlying logic into a bit of a maze.  This goals for this PR are to:

- Document a recipe for users to create custom accessors.  This now boils down to "subclass PandasDelegate".
- Unify the ways in which these accessors are constructed, including clarifying naming conventions.
- Keep never-directly-used methods like `_make_dt_accessor`, `_make_cat_accessor` out of the `Series` namespace.
- Avoid injecting names into existing namespaces.

The biggest simplification (and the main change in logic) is in where the `_make_foo_accessor` is defined.  Under the status quo, this looks roughly like:

```
    def _make_foo_acessor(self):
        [validation that self is foo-like]
        return FooDelegate(self)

    foo = AccessorProperty(FooDelegate, _make_foo_acessor)
```

This PR changes that, so that `_make_foo_accessor` is now a classmethod `_make_accessor` of `FooDelegate`.  Then in `Series` all we need is `foo = AccessorProperty(FooDelegate)`.

At the moment the documentation/recipe is in the form of an example in the module docstring to `core.accessors`.  Not sure where that belongs.  I'm not wild about the going back-and-forth between code that is copy/paste-able and ">>>" interludes.  `test_accessors` is a bare-bones copy of that example with some assertions.

 - [ ] closes #xxxx
 - [ ] tests added / passed
 - [x] passes ``git diff upstream/master -u -- "*.py" | flake8 --diff``
 - [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/accessors.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+        return property(fget=_getter, fset=_setter, doc=_doc)
+
+    @staticmethod
+    def create_delegator_method(name, delegate):
+        # Note: we really only need the `delegate` here for the docstring
+
+        def func(self, *args, **kwargs):
+            return self._delegate_method(name, *args, **kwargs)
+
+        if callable(name):
```
Comment (jbrockmendel) [inline pandas/core/accessors.py]: Note to self: you tried this option and decided against it.

File: pandas/core/accessors.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+                [...]
+
+
+        This replaces the older usage in which following a class definition
+        we would use `Foo._add_delegate_accessors(...)`.  The motivation
+        is that we would like to keep as much of a class's internals inside
+        the class definition.  For things that we cannot keep directly
+        in the class definition, a decorator is more directly tied to
+        the definition than a method call outside the definition.
+
```
Comment (jbrockmendel) [inline pandas/core/accessors.py]: The paragraph above may be helpful for the dev discussion, but probably doesn't belong in the docstring.

File: pandas/core/accessors.py [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+                overwrite the method/property in the target class if it exists
+            """
+            for name in accessors:
+                if typ == "property":
+                    func = Delegator.create_delegator_property(name, delegate)
+                else:
+                    func = Delegator.create_delegator_method(name, delegate)
+
+                # Allow for a callable to be passed instead of a name.
+                title = com._get_callable_name(name)
```
Comment (jbrockmendel) [inline pandas/core/accessors.py]: Note to self: you decided against this option.

File: pandas/core/frame.py [hunk 4]
```diff
@@ -6006,8 +6005,8 @@ def _put_str(s, space):
 
 # ----------------------------------------------------------------------
 # Add plotting methods to DataFrame
-DataFrame.plot = base.AccessorProperty(gfx.FramePlotMethods,
```
Comment (jbrockmendel) [inline pandas/core/frame.py]: Needs rebasing.

File: pandas/core/indexes/accessors.py [hunk 5]
```diff
@@ -228,18 +237,20 @@ class PeriodProperties(Properties):
...
+    def _make_accessor(cls, values):
+        try:
+            return maybe_to_datetimelike(values)
+        except Exception:
+            msg = "Can only use .dt accessor with datetimelike values"
+            raise AttributeError(msg)
+
+
+DatetimeAccessor = CombinedDatetimelikeProperties
+# Alias to mirror CategoricalAccessor
```
Comment (jbrockmendel) [inline pandas/core/indexes/accessors.py]: If we're subclassing `PandasDelegate`, would it make more sense for these to be named e.g. `DatetimeDelegate`?

File: pandas/core/strings.py [hunk 6]
```diff
@@ -1890,18 +1912,22 @@ def rindex(self, sub, start=0, end=None):
                                docstring=_shared_docs['ismethods'] %
                                _shared_docs['isdecimal'])
 
+    # TODO: Use this instead of wrapping all of these methods individually?
+    def _delegate_method(self, name, *args, **kwargs):
+        # TODO: It would be really nice to keep the signatures
+        method = getattr(self.values, name)
+        res = method(*args, **kwargs)
+        # TODO: Should this get wrapped in an index?
+        return res
```
Comment (jbrockmendel) [inline pandas/core/strings.py]: TODO: even though it isn't explicitly needed, having `StringMethods` subclass `PandasDelegate` will clarify its parallels with the other accessors.

Plus actually *using* `_delegate_method` could get rid of a lot of boilerplate.

File: pandas/core/strings.py [hunk 7]
```diff
@@ -1910,23 +1936,34 @@ def _make_str_accessor(self):
...
+
+
+StringAccessor = StringMethods  # Alias to mirror CategoricalAccessor
+
+
+# TODO: This is only mixed in to Index (this PR takes it out of Series)
+# and the _dir_additions/_dir_deletions won't play nicely with
+# any other class this gets mixed into that *does* implement its own
+# _dir_additions/_dir_deletions.  This should be deprecated.
+class StringAccessorMixin(object):
```
Comment (jbrockmendel) [inline pandas/core/strings.py]: `StringAccessorMixing` should be considered for deprecation.  See comment above.

File: pandas/core/indexes/category.py [hunk 8]
```diff
@@ -56,6 +71,11 @@ class CategoricalIndex(Index, base.PandasDelegate):
     _engine_type = libindex.Int64Engine
     _attributes = ['name']
 
+    def __init__(self, *args, **kwargs):
+        # Override to prevent accessors.PandasDelegate.__init__ from executing
+        # This is a kludge.
+        pass
```
Comment (jbrockmendel) [inline pandas/core/indexes/category.py]: Note to self: fix this.

File: pandas/core/indexes/accessors.py [hunk 9]
```diff
@@ -104,6 +106,9 @@ def _delegate_property_get(self, name):
                 result = result.astype('int64')
         elif not is_list_like(result):
             return result
+        elif isinstance(result, DataFrame):
+            # e.g. TimedeltaProperties.components
+            return result.set_index(self.index)
```
Comment (jbrockmendel) [inline pandas/core/indexes/accessors.py]: This allows `_delegate_property_get` to correctly handle `components`.

File: pandas/core/accessors.py [hunk 10]
```diff
@@ -0,0 +1,345 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+
+An example/recipe for creating a custom accessor.
```
Comment (jbrockmendel) [inline pandas/core/accessors.py]: @shoyer: You commented on this in #14781.  Have you found that custom accessors are frequently used by xarray users?

@jreback I can delete it or move it somewhere out of the way.  The `test_accessors` file might make a good place to park the discussion seeing as how it uses the same example anyway.

Comment (chris-b1) [inline pandas/core/accessors.py]: I'm +1 on the general idea of exposing a way to register custom accessors, though the api in your example seems a little clunky compared to what xarray does?  That said, may be better to punt that piece to a follow-up PR.

Comment (chris-b1) [inline pandas/core/accessors.py]: Agree that docs should live in `internals.rts`, logical companion to the subclassing docs

Comment (jbrockmendel) [inline pandas/core/accessors.py]: > example seems a little clunky compared to what xarray does

The examples are for pretty different use cases.  The xarray example defines a `center` attribute in terms of multiple existing [column]s of a [DataFrame].  That's why it doesn't need any vectorization boilerplate.

The example I used vectorizes properties/methods of the elements in a Series.  `StringMethods` or `CategoricalAccessor` work roughly this way.  (`CombinedDatetimelikeProperties` doesn't need to apply the property/method point-wise because they already exist in the underlying Series/Index.)

That said, if this isn't already obvious, that is a shortcoming of the documentation.

There is some avoidable clunkiness in that we are requiring every accessible attribute to be explicitly listed.

Comment (jbrockmendel) [inline pandas/core/accessors.py]: Just pushed commits that address some of these.  Took the example out of the module docstring, put a comment in PandasDelegate.__doc__ pointing readers towards the existing examples.

**Comment by shoyer** [comment]
This interface still looks really complicated to me, much more complex than [the recipe we have in xarray](http://xarray.pydata.org/en/stable/internals.html#extending-xarray). Based on the xarray recipe, you could write something like:
```python
@pd.register_series_accessor('my_accessor')
class MyAccessor(object):
    def __init__(self, series):
        ...
```
and voila, `series.my_accessor` returns a `MyAccessor` object.

In your case, library writers need to define lots of special methods like `_make_accessor`, `_delegate_property_get`, etc., not to mention the decorator that requires explicitly enumerating every method.

Why is all this necessary? This sort of complexity is OK internally, but not desirable for user facing API.

One reason I can think of is that we don't even define accessors if they aren't valid (e.g., `.cat` only exists on Series with a categorical dtype). But even here, I think this could require one extra method at most, e.g.,
```python
@pd.register_series_accessor('cat')
class CategoricalAccessor(object):
    def __init__(self, series):
        ...
    @classmethod
    def _should_exist(cls, series):
        # it might also work to just raise an error in __init__
        # in case where the accessor is invalid
        return series.dtype.is_categorical()
```

**Comment by jbrockmendel** [comment]
> Why is all this necessary? This sort of complexity is OK internally

@shoyer The first and most accurate answer is that it isn't, but there was a limit to how much legacy logic I wanted to change in my first non-trivial PR here.

Second, be sure that you are comparing analogous functions.  `xr.register_dataset_accessor('geo')(GeoAccessor)` corresponds roughly to `str = accessors.AccessorProperty(StringAccessor)`.

Using `_make_accessor` instead of just putting the validation in `__init__` is there for the branching cases in `pd.core.indexes.accessors`.  I mainly focused on getting `_make_dt_accessor` and `_make_cat_accessor` out of the `Series` namespace.  Now that I think about it, checking for `_make_accessor` and then falling back to `__init__` might be simpler.

I haven't looked through the history, but I expect that explicitly listing the delegated names was motivated by the desire to surface some properties/methods `DatetimeIndex` without inheriting all of it.

**Comment by shoyer** [comment]
> The first and most accurate answer is that it isn't, but there was a limit to how much legacy logic I wanted to change in my first non-trivial PR here.

This makes sense. I'll leave it up to others to judge if this internal clean-up is worth it, but let's refrain from adding any new public API until we have that cleaner interface.

**Comment by jbrockmendel** [comment]
If this is a sticking point, I'd argue that there are two refactorings here that are much more important than everything else.

1) `AccessorProperty.__init__` currently takes arguments `accessor_cls` and `construct_accesor`.  That constructor can often just be `accessor_cls.__init__`, but should *always* be method of the `accessor_cls`.  The `Series` (and `StringAccessorMixin`) namespace is definitely not where these belong.

2) `PandasDelegate._add_delegate_accessors` took a while to figure out.  De-nesting that method could save the next guy a headache.

**Comment by jbrockmendel** [comment]
A note transplanted from #17061: `StringAccessorMixin` can/should be deprecated.  Doing so requires only a small fraction of this PR.

Then `Index` can have `_accessor`, `_dir_deletions`, and `dir_additions` copy/pasted from Series (or both could inherit from something more generic).

@jreback Would you prefer 17061-style discussion go somewhere else like the mailing list?

**Comment by jreback** [comment]
u can certainly try the ML

File: pandas/core/accessors.py [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+        Parameters
+        ----------
+        data : the underlying object being accessed, usually Series or Index
+
+        Returns
+        -------
+        Delegate : instance of PandasDelegate or subclass
+
+        """
+        raise NotImplementedError(
```
Comment (jreback) [inline pandas/core/accessors.py]: use AbstractMethodError

File: pandas/core/accessors.py [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+
+        @delegate_names(Categorical, ["categories", "ordered"], "property")
+        class CategoricalAccessor(PandasDelegate):
+
+            @classmethod
+            def _make_accessor(cls, data):
+                [...]
+
+
+        This replaces the older usage in which following a class definition
```
Comment (jreback) [inline pandas/core/accessors.py]: you don't need to say what is *was*, just what it does now.

File: pandas/core/accessors.py [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+
+            Parameters
+            ----------
+            cls : the class to add the methods/properties to
+            delegate : the class to get methods/properties & doc-strings
+            acccessors : string list of accessors to add
+            typ : 'property' or 'method'
+            overwrite : boolean, default False
+                overwrite the method/property in the target class if it exists
+            """
```
Comment (jreback) [inline pandas/core/accessors.py]: maybe some asserts to valid types of the accessors and such

File: pandas/core/categorical.py [hunk 14]
```diff
@@ -2013,7 +2014,20 @@ def repeat(self, repeats, *args, **kwargs):
 # The Series.cat accessor
 
 
-class CategoricalAccessor(PandasDelegate, NoNewAttributesMixin):
+@accessors.wrap_delegate_names(delegate=Categorical,
```
Comment (jreback) [inline pandas/core/categorical.py]: an alternative way to do this is to add in the classes themselves which methods / properties should be considered accessors (either in a class method or via decorators), i did do this for a while, but actually though it simpler here. but can revisit at a later point.

File: pandas/core/categorical.py [hunk 15]
```diff
@@ -2047,31 +2068,23 @@ def _delegate_property_get(self, name):
...
-        return Series(self.categorical.codes, index=self.index)
-
     def _delegate_method(self, name, *args, **kwargs):
         from pandas import Series
         method = getattr(self.categorical, name)
         res = method(*args, **kwargs)
         if res is not None:
             return Series(res, index=self.index)
 
+    # TODO: Can we get this from _delegate_property_get?
```
Comment (jreback) [inline pandas/core/categorical.py]: yes, but its simpler here as it

File: pandas/core/indexes/accessors.py (hunk 5 above)
Comment (jreback) [inline pandas/core/indexes/accessors.py]: yes

File: pandas/core/series.py [hunk 16]
```diff
@@ -47,34 +56,25 @@
...
-                               Float64Index, _ensure_index)
-from pandas.core.indexing import check_bool_indexer, maybe_convert_indices
-from pandas.core import generic, base
+from pandas.core.config import get_option
 from pandas.core.internals import SingleBlockManager
 from pandas.core.categorical import Categorical, CategoricalAccessor
-import pandas.core.strings as strings
-from pandas.core.indexes.accessors import (
-    maybe_to_datetimelike, CombinedDatetimelikeProperties)
+
```
Comment (jreback) [inline pandas/core/series.py]: ideally don't move imports around unless they are germane, it makes the diff hard to read

File: pandas/core/strings.py [hunk 17]
```diff
@@ -1316,6 +1358,27 @@ def str_encode(arr, encoding, errors="strict"):
     return _na_map(f, arr)
 
 
+def str_normalize(arr, form):
```
Comment (jreback) [inline pandas/core/strings.py]: ?

File: pandas/core/strings.py (hunk 7 above)
Comment (jreback) [inline pandas/core/strings.py]: this is internal, you can just remove it if not needed

**Review by jreback — COMMENTED** [review]
generally looks ok. lmk when you are ready ping.

File: pandas/core/series.py (hunk 16 above)
Comment (jbrockmendel) [inline pandas/core/series.py]: Noted.

File: pandas/core/indexes/accessors.py (hunk 5 above)
Comment (jbrockmendel) [inline pandas/core/indexes/accessors.py]: Changing these names over now.

File: pandas/core/categorical.py (hunk 14 above)
Comment (jbrockmendel) [inline pandas/core/categorical.py]: I'm amenable to that.  Over in `core.indexes.accessors` there is a comment above `DatetimeAccessor` noting an alternative way to define the accessed properties/methods.

The only pattern I actively dislike is `CategoricalIndex._add_accessors`; pinning attributes to the class from a distance can lead to surprises (like the still-mysterious circular import `core.series`<-->`plotting._core`).

File: pandas/core/strings.py (hunk 17 above)
Comment (jbrockmendel) [inline pandas/core/strings.py]: Nearly all the StringMethods methods follow a pattern of defining `str_foo` as a module-level function and then wrapping it as the `foo` method.  `StringMethods.normalize` didn't follow this pattern, so I moved it out.  At the time I intended to use `_delegate_method` to get rid of a bunch of boilerplate in the class (and clarify its relation to other `PandasDelegate` subclasses), but eventually chose to hold off on that.

So this isn't strictly pertinent to this PR; I'll be happy to revert it.

File: pandas/core/strings.py (hunk 7 above)
Comment (jbrockmendel) [inline pandas/core/strings.py]: Great; wasn't sure if it should be saved for a separate PR.  Doing so requires copy/pasting `_accessors`, `_dir_deletions`, and `_dir_additions` from `Series` (which I see as a good thing, per #8162 , #17061)

**Comment by jbrockmendel** [comment]
Does the CircleCI timeout require action?

**Comment by chris-b1** [comment]
I restarted it; think it's been a bit flaky

**Comment by gfyoung** [comment]
> I restarted it; think it's been a bit flaky

Yeah, the CI services can sometimes be annoying like.  Travis I've found is generally the least "disobedient" (for lack of a better term), but it has its own share of problems :cry:

File: pandas/core/strings.py (hunk 17 above)
Comment (jbrockmendel) [inline pandas/core/strings.py]: Would getting rid of the not-strictly-relevant edits here in strings.py grease the wheels on this?

**Comment by jreback** [comment]
you can rebase or close this. prob easier to do step by step changes.

**Comment by jbrockmendel** [comment]
Do you mind letting this stay open until #17117 is resolved?  I'll have to rebase again after that anyway.
